canonical-web-and-design/maas-ui
# [Subnets] Add search

## Done

-  Add search to Subnets table
  - refactor grouping - separate table data from visual representation
    - grouping is now done after pagination
    - this is to make sure that the first item from a group on each page always has a description (in contrary to how this works on machines listing)
    - add `clone-deep` to enable easy object cloning (1.3kB minified and gzipped https://bundlephobia.com/package/clone-deep@4.0.1)
  - remove redundant react-table types (now using `ModularTable` types from `react-components`)
  
## Performance impact
Measurements below are averages from 3 consecutive runs of visiting `/networks?by=fabric page` with 145 subnets table rows

| metric | master |  this branch | improvement
|--------|-------|---|---|
| Lighthouse performance score | 66 | 77  | **+11 points** |
| SubnetsTable render |    43ms   |  18ms | **2.4x faster** | 
| from loaded* to calculated rows |    77ms   |  21ms  | **3.5x faster**  |

* [loaded = fabricsLoaded && vlansLoaded && subnetsLoaded && spacesLoaded](https://github.com/petermakowski/maas-ui/blob/cf230777cd7c331fa3953f506a34b025c24522a7/ui/src/app/subnets/views/SubnetsList/SubnetsTable/hooks.ts#L30)

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to `/networks` and try various search queries
- Verify that correct rows are displayed
- Try adding/removing items in another tab and verify the table behaves as expected (it generally tries to preserve the currently displayed page index unless the page is out of bounds after updates coming from the backend)

## Fixes

https://github.com/canonical-web-and-design/app-tribe/issues/759

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against *master* rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in master.
